rfcs/20200624-pluggable-device-for-tensorflow/tutorial.md

```diff
@@ -0,0 +1,1000 @@
+* `params->profiler_fns->struct_size`: plugin needs to set it as `TP_PROFILER_FNS_STRUCT_SIZE` (defined in pluggable_profiler.h). This field is used for the Profiler C API version check between TensorFlow and the plugin.
+* `params->profiler->type`: This field is the device type(e.g. GPU, APU..).
+* `params->profiler_fns->start`: a callback for starting the profiler.
+```c++
+void profiler_start(const TP_Profiler* profiler, TF_Status* status) {
+  /* Enable profiler */
+  ...
+}
+```
+* `params->profiler_fns->stop`: a callback for stopping the profiler.
+```c++
+void profiler_stop(const TP_Profiler* profiler, TF_Status* status) {
+  /* Disable Profiler */
+  ...
+}
+```
+* `params->profiler_fns->collect_data_xspace`: a callback for saving collected profile data into XSpace and serializers it into the buffer. If this have been called, subsequent calls might return empty data.
+```c++
+void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t*
+buffer, size_t* size_in_bytes, TF_Status* status) {
+  Xspace xspace = get_my_xspace(); /* Plugin generates Xspace based on
+  collected profiler data. */ size_t buffer_size_in_bytes = *size_in_bytes;
+  *size_in_bytes = xspace.ByteSizeLong(); /* get the size of Xspace */
+  if (buffer == nullptr) {
+    return; /* TensorFlow will first get the size of Xspace, then allocate
+    the big enough buffer and pass it to the plugin for retrieving Xspace.
+    */
+  }
+  bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
+}
+
+```
+* `params->destroy_profiler`: pointer to plugin's `TP_Profiler` clean up function. Cleans up fields inside `TP_Profiler` that were allocated by the plugin. `profiler` itself must not be deleted by the plugin.
+* `params->destroy_profiler_fns`: pointer to plugin's `TP_ProfilerFns` clean up function. Cleans up fields inside `TP_ProfilerFns` that were allocated by the plugin. `profiler_fns` itself must not be deleted by the plugin.
+
+you can find profiler implementation sample code in `sample/tensorflow_plugin/src/profiler`.
+* profiler example:
+ <div>
+<img src=profiler_result.png>
+</div>
+
+## **Plugin build**
+
+After implementing the plugin, we need to build it as a dynamic library. Build system is decided by plugin authors, you can choose bazel, cmake or other build systems, it is out of scope in this tutorial. To make things simple, we just use the gcc command here.
+
+When building the plugin, we have two dependencies here:
+
+1. We need to include those C API header files provided by Core TensorFlow.
+
+2. The built plugin library needs to add dependency to `_pywrap_tensorflow_internal.so`, which is built by Core TensorFlow. `_pywrap_tensorflow_internal.so` contains those C API implementations. If you don’t add this dependency, it will report an "undefined symbol" error when loading the plugin library.
```

Comment:
Working on my Rust port I found that linking to `libtensorflow_framework.so.2` works fine as well, but does not add a dependency on Python, which is not necessary and limits the possibilities. I changed those libraries when I found that tests do not build because of linking dependency on Python's C APIs, but with C-native lib everything is fine.

See https://github.com/sh7dm/rust-tf-pluggabledevice/commit/d1d812c337f12644f97c35e8038e7d79341cd9e5